### PR TITLE
KAA-308:  C EP SDK abstraction layer

### PR DESCRIPTION
**Pull request by andreypanasenko**
Reviewers: 
https://github.com/akokhanovskyi 
https://github.com/yzeygerman

### CODE REVIEW

**Comment by ashvayka** [comment]
Please update PR name to match our pattern (issue number + subject from Jira)

**Comment by izeigerman** [comment]
LGTM

**Comment by ghost** [comment]
LGTM

File: client/client-multi/client-c/src/gen/kaa_profile_gen.c [hunk 1]
```diff
@@ -64,7 +61,7 @@ static size_t kaa_profile_basic_endpoint_profile_test_get_size(void* data)
 
 kaa_profile_basic_endpoint_profile_test_t* kaa_profile_basic_endpoint_profile_test_create()
 {
-    kaa_profile_basic_endpoint_profile_test_t* record =
+    kaa_profile_basic_endpoint_profile_test_t* record =
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/gen/kaa_profile_gen.c:64]: What is this trailing space for? :)

File: client/client-multi/client-c/src/kaa_common_schema.c [hunk 2]
```diff
@@ -14,10 +14,13 @@
  * limitations under the License.
  */
 
+#include <string.h>
+#include "platform/stdio.h"
+#include "kaa_common_schema.h"
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_common_schema.c:19]: Why do you #include "kaa_common_schema.h" twice here?

File: client/client-multi/client-c/src/kaa_context.h [hunk 3]
```diff
@@ -28,23 +28,52 @@
 extern "C" {
 #endif
 
-typedef struct kaa_status_t             kaa_status_t;
-typedef struct kaa_platform_protocol_t  kaa_platform_protocol_t;
+#ifndef KAA_STATUS_T
+    #define KAA_STATUS_T
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_context.h]: Let us please have hash at the beginning of the line even for indented preprocessor commands:
#ifndef KAA_STATUS_T
#   define KAA_STATUS_T

File: client/client-multi/client-c/src/kaa_status.c [hunk 4]
```diff
@@ -13,7 +13,11 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "platform/stdio.h"
+#include "platform/sha.h"
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_status.c]: Are you sure you absolutely need every of these includes here? Let us please not include anything that is not absolutely necessary.

File: client/client-multi/client-c/src/platform-impl/Econais/EC19D/econais_ec19d_platform.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ econais_ec19d_platform.h
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/Econais/EC19D/econais_ec19d_platform.h:18]: Please replace Eclipse auto-generated docs with doxygen-compatible docs (use @file; @brief, etc.)

File: client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    }
+    sndc_printf(buffer);
+}
+
+time_t get_systime()
+{
+    return (time_t) sndc_sys_getTimestamp_msec();
+}
+
+int kaa_format_sprintf(char * buffer, size_t buffer_size, const char * FORMAT,
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c]: These functions are not core part of Kaa SDK. Their names should start with "ext_".

File: client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+#include "../../../platform/system_logger.h"
+
+void write_log(FILE * sink, const char * buffer, size_t message_size)
+{
+    if (!buffer) {
+        return;
+    }
+    sndc_printf(buffer);
+}
+
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c:30]: Per code style, for readability, 3 empty lines must separate functions and other unrelated code blocks (includes, type declarations, etc.)

File: client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        memset(p_format,0,p_format_size); \
+        strncpy(p_format,last_format,formated_size);\
+        strncpy(p_format + formated_size, symbols, symbols_size); \
+        _INT_SPRINTF(type)
+
+int kaa_logger_sprintf(char * buffer, size_t buffer_size, const char * format, va_list args)
+{
+    int format_length = strnlen(format, buffer_size);
+    int i = 0, res_len = 0;
+    const char * zu_symbols = "%lu";
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform-impl/Econais/EC19D/logger.c]: No need in space after "*":
const char *zu_symbols ...
Similarly going forward.

File: client/client-multi/client-c/src/platform/system_logger.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * limitations under the License.
+ */
+
+#ifndef SYSTEM_LOGGER_H_
+#define SYSTEM_LOGGER_H_
+
+#include "../kaa_error.h"
+
+/*
+ * Write log message in system dependent way.
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform/system_logger.h]: This documentation is insufficient. Please take a look at the other headers where functions are fully documented in doxygen-compatible fashion, including @brief, @param, and @return at least.

File: client/client-multi/client-c/src/platform/system_logger.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+ * Return current system time
+ */
+time_t get_systime();
+
+/**
+ * Put formated LOG prefix in buffer
+ * LOG prefix format example:
+ *      1970/01/01 2:30:36 [TRACE] [kaa_bootstrap.c:38] (0) -
+ */
+int kaa_format_sprintf(char * buffer, size_t buffer_size, const char * FORMAT, const char * LOG_LEVEL_NAME, const char * truncated_name, int lineno, kaa_error_t error_code);
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform/system_logger.h]: Why are the parameter names in CAPS? These are not macros.

Comment (akokhanovskyi) [inline client/client-multi/client-c/src/platform/system_logger.h]: The function names must start with "ext_" for everything that is not core part of Kaa SDK.

File: client/client-multi/client-c/src/utilities/kaa_log.c [hunk 11]
```diff
@@ -124,45 +123,40 @@ void kaa_log_write(kaa_logger_t *self, const char* source_file, int lineno, kaa_
...
         consumed_len += res_len;
-        if (consumed_len > self->buffer_size - 1) {
+        if (consumed_len > self->buffer_size - 2) {
             // Ran out of buffer space
-            consumed_len = self->buffer_size - 1;
+            consumed_len = self->buffer_size - 2;
         }
     }
 
     // Terminate buffer with '\n'. Null-termination is not important because the buffer length is specified in fwrite() below
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/utilities/kaa_log.c]: Fix the comment here. You terminate string with "\n\0", not with "\n"

File: client/client-multi/client-c/test/kaa_test_external.c [hunk 12]
```diff
@@ -17,6 +17,8 @@
 #include "kaa_external.h"
 
 #include <string.h>
+#include <stdio.h>
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/test/kaa_test_external.c:20]: Why do you need these new includes here?

File: client/client-multi/client-c/test/test_kaa_bootstrap.c [hunk 13]
```diff
@@ -15,6 +15,10 @@
  */
 
 #include <string.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
```
Comment (akokhanovskyi) [inline client/client-multi/client-c/test/test_kaa_bootstrap.c]: Why do you need these new includes here?

File: client/client-multi/client-c/src/kaa_status.c (hunk 4 above)
Comment (andreypanasenko) [inline client/client-multi/client-c/src/kaa_status.c]: I'm absolutely sure, I've checked every C file ob both linux and econais platforms.

File: client/client-multi/client-c/test/test_kaa_bootstrap.c (hunk 13 above)
Comment (andreypanasenko) [inline client/client-multi/client-c/test/test_kaa_bootstrap.c]: Because without this include its not compile.
We remove all system dependent definitions from headers, so they should be in every compiled C files.

File: client/client-multi/client-c/src/kaa_status.c (hunk 4 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/src/kaa_status.c]: Okay. Just checking :)

File: client/client-multi/client-c/test/test_kaa_bootstrap.c (hunk 13 above)
Comment (akokhanovskyi) [inline client/client-multi/client-c/test/test_kaa_bootstrap.c]: Okay, makes sense.
